File: packages/server-go/internal/bot/actions.go
package bot

import (
	"context"
	"strings"

	"github.com/tgs/server/internal/bus"
	"github.com/tgs/server/internal/domain"
	"github.com/tgs/server/internal/rules"
	"github.com/tgs/server/internal/sanction"
	"github.com/tgs/server/internal/store"
	"github.com/tgs/server/internal/tgapi"
)

// 规则命中的处置执行。
//
// 与 rules 包的分工：引擎判断「命中了什么」，这里决定「因此做什么」。
// 分开的硬性理由：面板的正则沙盒要复用引擎，但绝不能因为管理员点了
// 「测试」就真的去删消息或封禁用户。

// hitContext 是一次处置的上下文。
type hitContext struct {
	settings        store.BotSettings
	contact         store.ContactRow
	scope           string // user | admin
	sourceChatID    int64
	sourceMessageID int64
	topicID         *int64
	threadID        *int64
	messageRowID    *int64
	text            string
}

// actionReport 是一次处置的结果。
type actionReport struct {
	Outcomes  []string
	Blocked   bool
	HitIDs    []int64
	Violation *sanction.Outcome
	Banned    bool
}

// applyRuleHits 执行一次命中处置。
//
// 顺序有讲究：**先删原消息，再发警告**。反过来会让用户在「收到警告」到
// 「消息被删」之间有一个窗口去截图或转发 —— 这是真实存在的对抗场景。
func (r *Runtime) applyRuleHits(ctx context.Context, hc hitContext, hits []rules.Hit) actionReport {
	report := actionReport{Outcomes: make([]string, 0, 4)}

	if len(hits) == 0 {
		return report
	}

	// 管理员的消息只审计不处罚：规则是给终端用户定的，
	// 让管理员被自己的规则静默掉是荒谬的。
	if hc.scope == "admin" {
		report.Outcomes = append(report.Outcomes, domain.OutcomeLoggedOnly)
		r.recordHits(ctx, hc, hits, report.Outcomes, &report)
		return report
	}

	actions := rules.UnionActions(hits)

	// 1. 删除用户私聊里的原消息
	if actions[domain.ActionDelete] {
		deleted := r.deleteOrigin(ctx, hc)
		report.Outcomes = append(report.Outcomes, boolOutcome(deleted, domain.OutcomeDeleted, domain.OutcomeDeleteFailed))
	}

	// 2. 阶梯处罚（仅 escalate 动作触发）
	if actions[domain.ActionEscalate] {
		// 取最高分而不是求和：三条轻规则叠成一次重罚会让阈值失去意义
		severity := rules.TotalSeverity(hits)
		var ruleID *int64
		if hits[0].Rule.ID != 0 {
			id := hits[0].Rule.ID
			ruleID = &id
		}

		outcome, err := r.sanction.Apply(ctx, sanction.ViolationInput{
			BotID:     r.botID,
			ContactID: hc.contact.ID,
			RuleID:    ruleID,
			Severity:  severity,
			Reason:    domain.ReasonRuleHit,
		})
		if err != nil {
			r.log.Error("阶梯处罚失败", "contactId", hc.contact.ID, "err", err)
			report.Outcomes = append(report.Outcomes, domain.OutcomeLoggedOnly)
		} else {
			report.Violation = &outcome
			if outcome.Escalated && outcome.CurrentType != nil {
				report.Outcomes = append(report.Outcomes, outcomeForSanction(*outcome.CurrentType))
			} else {
				report.Outcomes = append(report.Outcomes, domain.OutcomeLoggedOnly)
			}
		}
	}

	// 3. 用户侧通知：阶梯升级优先，避免一次命中发两条
	if report.Violation != nil && report.Violation.Escalated {
		if r.notifyUserOfSanction(ctx, hc, hits, report.Violation) {
			report.Outcomes = append(report.Outcomes, domain.OutcomeWarned)
		}
	} else if actions[domain.ActionWarn] {
		text := renderTemplate(hc.settings.WarnTemplate, map[string]string{
			"name":     hc.contact.DisplayName(),
			"username": deref(hc.contact.Username),
			"id":       itoa(hc.contact.TgUserID),
			"ruleName": hits[0].Rule.Name,
			"matched":  hits[0].MatchedText,
			"score":    itoa(int64(hc.contact.ViolationScore)),
			"botName":  r.botName,
		})
		if sent, _ := r.sendToUser(ctx, hc.contact.TgUserID, text); sent {
			report.Outcomes = append(report.Outcomes, domain.OutcomeWarned)
		}
	}

	// 4. 管理员侧告警卡片
	shouldNotify := hc.settings.NotifyAdmins && hc.threadID != nil &&
		(actions[domain.ActionNotify] ||
			(report.Violation != nil && report.Violation.Escalated &&
				report.Violation.Step != nil && report.Violation.Step.NotifyAdmins))

	if shouldNotify {
		if r.sendAlertCard(ctx, hc, hits, report.Outcomes, report.Violation) {
			report.Outcomes = append(report.Outcomes, domain.OutcomeNotified)
		} else {
			report.Outcomes = append(report.Outcomes, domain.OutcomeNotifyFailed)
		}
	}

	// 5. 落审计：每条命中一行（「每条都记，动作取并集」）
	r.recordHits(ctx, hc, hits, report.Outcomes, &report)

	// 6. 拉黑后关闭话题。刻意「关闭」而不是「删除」：删除会丢掉全部历史，
	//    而拉黑往往正是最需要保留证据的时候。
	if report.Violation != nil && report.Violation.CurrentType != nil &&
		*report.Violation.CurrentType == domain.SanctionBan &&
		hc.topicID != nil && hc.threadID != nil {
		report.Banned = true
		if err := r.closeTopic(ctx, *hc.topicID, *hc.threadID); err != nil {
			r.log.Warn("关闭话题失败", "topicId", *hc.topicID, "err", err)
		}
	}

	// 7. 撤回已转发进话题的副本（先转发、后新增规则命中的场景）
	if hc.messageRowID != nil && actions[domain.ActionDelete] {
		r.deleteMirror(ctx, *hc.messageRowID)
	}

	report.Blocked = report.Violation != nil && report.Violation.CurrentType != nil &&
		isBlockedTier(*report.Violation.CurrentType)

	return report
}

// recordHits 写入命中审计并自增计数。
func (r *Runtime) recordHits(
	ctx context.Context,
	hc hitContext,
	hits []rules.Hit,
	outcomes []string,
	report *actionReport,
) {
	ids := make([]int64, 0, len(hits))

	for _, hit := range hits {
		var ruleID *int64
		if hit.Rule.ID != 0 {
			id := hit.Rule.ID
			ruleID = &id
		}
		matched := hit.MatchedText
		if matched == "" {
			matched = truncate(hc.text, 200)
		}

		var excerpt *string
		if hit.NormalizedExcerpt != "" {
			excerpt = &hit.NormalizedExcerpt
		}

		hitID, err := r.db.RecordRuleHit(ctx, store.RecordRuleHitInput{
			RuleID:            ruleID,
			RuleName:          hit.Rule.Name,
			RulePattern:       hit.Rule.Pattern,
			RuleFlags:         hit.Rule.Flags,
			BotID:             r.botID,
			ContactID:         hc.contact.ID,
			TopicID:           hc.topicID,
			MessageID:         hc.messageRowID,
			MatchedText:       matched,
			NormalizedExcerpt: excerpt,
			Outcomes:          outcomes,
			Severity:          hit.Severity,
		})
		if err != nil {
			r.log.Error("写入命中审计失败", "ruleId", hit.Rule.ID, "err", err)
			continue
		}
		ids = append(ids, hitID)

		r.bus.Publish(bus.ChannelGlobal, bus.EventRuleHit, domain.RuleHit{
			ID:                hitID,
			RuleID:            ruleID,
			RuleName:          hit.Rule.Name,
			RulePattern:       hit.Rule.Pattern,
			RuleFlags:         hit.Rule.Flags,
			BotID:             r.botID,
			BotName:           r.botName,
			ContactID:         hc.contact.ID,
			ContactName:       hc.contact.DisplayName(),
			ContactUsername:   hc.contact.Username,
			TgUserID:          hc.contact.TgUserID,
			TopicID:           hc.topicID,
			ThreadID:          hc.threadID,
			MatchedText:       matched,
			NormalizedExcerpt: excerpt,
			Outcomes:          outcomes,
			Severity:          hit.Severity,
			CreatedAt:         nowMillis(),
		})
	}

	if len(ids) > 0 {
		ruleIDs := make([]int64, 0, len(hits))
		for _, hit := range hits {
			if hit.Rule.ID != 0 {
				ruleIDs = append(ruleIDs, hit.Rule.ID)
			}
		}
		if err := r.engine.BumpHitCounts(ctx, ruleIDs); err != nil {
			r.log.Warn("更新命中计数失败", "err", err)
		}
	}
	report.HitIDs = ids
}

// deleteOrigin 撤回用户私聊里的原消息。
// Bot API 明确允许删除私聊中的 incoming 消息，但有 48 小时的时间窗。
func (r *Runtime) deleteOrigin(ctx context.Context, hc hitContext) bool {
	if !hc.settings.DeleteOriginMessage {
		return false
	}
	if err := r.api.DeleteMessage(ctx, hc.sourceChatID, hc.sourceMessageID); err != nil {
		if tgapi.IsTooOldToDelete(err) {
			r.log.Debug("删除原消息失败：超过 48 小时窗口", "messageId", hc.sourceMessageID)
		} else {
			r.log.Debug("删除原消息失败", "messageId", hc.sourceMessageID, "err", err)
		}
		return false
	}
	return true
}

// deleteMirror 撤回已转发进话题的副本。
//
// 用于「先转发、后新增规则命中」的场景：规则是在消息已经进了话题之后
// 才被加上或改动的，此时话题里那份副本需要补删。
//
// 失败只记日志：副本留着不影响正确性，而删不掉最常见的原因是
// 超过 Telegram 的 48 小时窗口 —— 那不是我们能解决的问题。
func (r *Runtime) deleteMirror(ctx context.Context, messageRowID int64) {
	mapped, err := r.db.GetMessage(ctx, messageRowID)
	if err != nil {
		r.log.Debug("查不到消息记录，跳过副本撤回", "messageRowId", messageRowID, "err", err)
		return
	}
	if mapped.DestChatID == nil || mapped.RelayedMessageID == nil {
		return
	}

	if err := r.api.DeleteMessage(ctx, *mapped.DestChatID, *mapped.RelayedMessageID); err != nil {
		r.log.Debug("撤回副本失败", "messageRowId", messageRowID, "err", err)
		return
	}
	if err := r.db.MarkMessageDeleted(ctx, mapped.ID); err != nil {
		r.log.Warn("标记副本已删除失败", "err", err)
		return
	}
	r.publishMessage(ctx, mapped, bus.EventMessageDeleted)
}

// notifyUserOfSanction 按阶梯档位给用户发提示文案。
func (r *Runtime) notifyUserOfSanction(
	ctx context.Context,
	hc hitContext,
	hits []rules.Hit,
	violation *sanction.Outcome,
) bool {
	if violation.Step == nil {
		return false
	}

	// 静默刻意不通知：它的全部意义就是让对方无感知。
	// 一发提示等于告诉对方「你的消息被吞了，换个号再来」。
	if violation.Step.Type == domain.SanctionSilence {
		return false
	}

	tmpl := hc.settings.WarnTemplate
	switch violation.Step.Type {
	case domain.SanctionMute:
		tmpl = hc.settings.MuteTemplate
	case domain.SanctionBan:
		tmpl = hc.settings.BanTemplate
	}

	var expiresAt *int64
	if violation.Sanction != nil {
		expiresAt = violation.Sanction.ExpiresAt
	}

	ruleName := ""
	if len(hits) > 0 {
		ruleName = hits[0].Rule.Name
	}

	text := renderTemplate(tmpl, map[string]string{
		"name":     hc.contact.DisplayName(),
		"username": deref(hc.contact.Username),
		"id":       itoa(hc.contact.TgUserID),
		"ruleName": ruleName,
		"score":    itoa(int64(violation.Score)),
		"until":    formatUntil(expiresAt, r.loc),
		"botName":  r.botName,
	})
	if text == "" {
		return false
	}

	sent, _ := r.sendToUser(ctx, hc.contact.TgUserID, text)
	return sent
}

// sendAlertCard 发送话题内的告警卡片。
func (r *Runtime) sendAlertCard(
	ctx context.Context,
	hc hitContext,
	hits []rules.Hit,
	outcomes []string,
	violation *sanction.Outcome,
) bool {
	if hc.threadID == nil {
		return false
	}

	names := make([]string, 0, len(hits))
	for _, hit := range hits {
		names = append(names, hit.Rule.Name)
	}

	matched := ""
	if len(hits) > 0 {
		matched = hits[0].MatchedText
	}

	score := hc.contact.ViolationScore
	sanctionType := ""
	if violation != nil {
		score = violation.Score
		if violation.CurrentType != nil {
			sanctionType = *violation.CurrentType
		}
	}

	note := ""
	if len(hits) > 0 && hits[0].Rule.Note != nil {
		note = *hits[0].Rule.Note
	}

	text := renderTemplate(hc.settings.AlertCardTemplate, map[string]string{
		"name":     hc.contact.DisplayName(),
		"username": orDefault(atPrefix(deref(hc.contact.Username)), "（无用户名）"),
		"id":       itoa(hc.contact.TgUserID),
		"ruleName": strings.Join(names, "、"),
		"matched":  matched,
		"outcome":  describeOutcomes(outcomes, sanctionType),
		"score":    itoa(int64(score)),
		"botName":  r.botName,
		"reason":   note,
	})

	id, err := r.sendToTopic(ctx, *hc.threadID, text, nil, false)
	return err == nil && id != 0
}

// describeOutcomes 把处置动作翻成一句面向管理员的人话。
func describeOutcomes(outcomes []string, sanctionType string) string {
	switch sanctionType {
	case domain.SanctionBan:
		return "拉黑"
	case domain.SanctionMute:
		return "禁言"
	case domain.SanctionSilence:
		return "静默"
	case domain.SanctionWarn:
		return "警告"
	}

	labels := map[string]string{
		domain.OutcomeDeleted:      "已删除",
		domain.OutcomeDeleteFailed: "删除失败",
		domain.OutcomeNotified:     "已通知管理员",
		domain.OutcomeNotifyFailed: "通知失败",
		domain.OutcomeWarned:       "已警告",
		domain.OutcomeLoggedOnly:   "仅记录",
		domain.OutcomeRegexTimeout: "规则超时",
	}

	seen := make(map[string]bool)
	parts := make([]string, 0, len(outcomes))
	for _, o := range outcomes {
		label, ok := labels[o]
		if !ok || seen[label] {
			continue
		}
		seen[label] = true
		parts = append(parts, label)
	}
	if len(parts) == 0 {
		return "仅记录"
	}
	return strings.Join(parts, " + ")
}

func outcomeForSanction(tier string) string {
	switch tier {
	case domain.SanctionWarn:
		return domain.OutcomeWarned
	case domain.SanctionSilence:
		return domain.OutcomeSilenced
	case domain.SanctionMute:
		return domain.OutcomeMuted
	case domain.SanctionBan:
		return domain.OutcomeBanned
	}
	return domain.OutcomeLoggedOnly
}

func isBlockedTier(tier string) bool {
	for _, t := range domain.BlockingSanctions {
		if t == tier {
			return true
		}
	}
	return false
}

func boolOutcome(v bool, whenTrue, whenFalse string) string {
	if v {
		return whenTrue
	}
	return whenFalse
}

File: packages/server/src/rules/actions.ts
import {
  channel,
  type BotSettings,
  type HitOutcome,
  type SanctionType,
} from '@tgs/shared';
import type { Api } from 'grammy';
import { eq } from 'drizzle-orm';
import { getDb } from '../db/client.ts';
import { messages, ruleHits } from '../db/schema.ts';
import { publish } from '../core/bus.ts';
import { logger } from '../core/logger.ts';
import { getGlobalSettings } from '../core/settings.ts';
import { formatUntil, renderTemplate } from '../core/template.ts';
import { recordViolation, type ViolationOutcome } from '../sanctions.ts';
import {
  closeTopic,
  sendToTopic,
  sendToUser,
  touchTopic,
  type ContactRow,
} from '../bots/topics.ts';
import { deleteMirror } from '../bots/relay.ts';
import {
  autoDisableRule,
  bumpHitCounts,
  unionActions,
  type RuleEvaluation,
  type RuleHit,
  type RuleScope,
} from './engine.ts';

/**
 * 规则命中的处置执行。
 *
 * 与 `engine.ts` 的分工：引擎判断「命中了什么」，这里决定「因此做什么」。
 * 分开有一条硬性理由：面板里的正则沙盒要复用引擎，但绝不能因为管理员
 * 点了「测试」就真的去删消息或封禁用户。
 */

export interface HitContext {
  api: Api;
  botId: number;
  botName: string;
  adminGroupId: number;
  settings: BotSettings;
  contact: ContactRow;
  /** 用户消息（可处罚）还是管理员消息（仅审计） */
  scope: RuleScope;
  /** 触发规则的原始消息所在的聊天与 id */
  sourceChatId: number;
  sourceMessageId: number;
  /** 落库的中继记录 id；命中即拦截时为 null */
  messageRowId: number | null;
  topicId: number | null;
  threadId: number | null;
  /** 被匹配的原文，用于 matchedText 兜底 */
  rawText: string;
}

export interface ActionReport {
  outcomes: HitOutcome[];
  /** 是否拦截这条消息（不转发） */
  blocked: boolean;
  hitIds: number[];
  violation: ViolationOutcome | null;
  /** 本次把用户拉黑了 —— 调用方据此关闭话题 */
  banned: boolean;
}

/**
 * 写一条命中审计。
 *
 * `rulePattern` / `ruleFlags` 是**快照**：规则随时可能被改甚至被删，
 * 而审计记录必须永远能回答「当时是按什么判定的」。
 */
async function insertHit(
  ctx: HitContext,
  hit: RuleHit,
  outcomes: HitOutcome[],
): Promise<number | null> {
  const { db } = getDb();
  const now = Date.now();
  try {
    const inserted = await db
      .insert(ruleHits)
      .values({
        ruleId: hit.rule.id,
        ruleName: hit.rule.name,
        rulePattern: hit.rule.pattern,
        ruleFlags: hit.rule.flags,
        botId: ctx.botId,
        contactId: ctx.contact.id,
        topicId: ctx.topicId,
        messageId: ctx.messageRowId,
        matchedText: hit.matchedText || ctx.rawText.slice(0, 200),
        normalizedExcerpt: hit.normalizedExcerpt,
        outcomes,
        severity: hit.severity,
        createdAt: now,
      })
      .returning({ id: ruleHits.id });

    const id = inserted[0]?.id ?? null;
    if (id !== null) {
      publish(channel.global, 'rule.hit', {
        id,
        ruleId: hit.rule.id,
        ruleName: hit.rule.name,
        rulePattern: hit.rule.pattern,
        ruleFlags: hit.rule.flags,
        botId: ctx.botId,
        botName: ctx.botName,
        contactId: ctx.contact.id,
        contactName:
          [ctx.contact.firstName, ctx.contact.lastName].filter(Boolean).join(' ') || '未知用户',
        contactUsername: ctx.contact.username,
        tgUserId: ctx.contact.tgUserId,
        topicId: ctx.topicId,
        threadId: ctx.threadId,
        matchedText: hit.matchedText,
        normalizedExcerpt: hit.normalizedExcerpt,
        outcomes,
        severity: hit.severity,
        createdAt: now,
      });
    }
    return id;
  } catch (err) {
    logger.error({ err, ruleId: hit.rule.id }, '写命中审计失败');
    return null;
  }
}

/** 撤回用户私聊里的原消息。Bot API 明确允许删除私聊中的 incoming 消息 */
async function deleteOrigin(ctx: HitContext): Promise<boolean> {
  if (!ctx.settings.deleteOriginMessage) return false;
  try {
    await ctx.api.deleteMessage(ctx.sourceChatId, ctx.sourceMessageId);
    return true;
  } catch (err) {
    logger.debug({ err }, '删除用户原消息失败（通常是超过 48 小时）');
    return false;
  }
}

/** 处置动作 → 面向管理员的简短描述，用于告警卡片里的 `{outcome}` */
function describeOutcomes(outcomes: HitOutcome[], sanctionType: SanctionType | null): string {
  if (sanctionType === 'ban') return '拉黑';
  if (sanctionType === 'mute') return '禁言';
  if (sanctionType === 'silence') return '静默';
  if (sanctionType === 'warn') return '警告';

  const labels: Partial<Record<HitOutcome, string>> = {
    deleted: '已删除',
    delete_failed: '删除失败',
    notified: '已通知管理员',
    notify_failed: '通知失败',
    warned: '已警告',
    logged_only: '仅记录',
    regex_timeout: '规则超时',
  };
  const named = outcomes.map((o) => labels[o]).filter((v): v is string => Boolean(v));
  return named.length > 0 ? [...new Set(named)].join(' + ') : '仅记录';
}

function outcomeForSanction(type: SanctionType): HitOutcome {
  switch (type) {
    case 'warn':
      return 'warned';
    case 'silence':
      return 'silenced';
    case 'mute':
      return 'muted';
    case 'ban':
      return 'banned';
  }
}

/**
 * 执行一次命中处置。
 *
 * 顺序有讲究：**先删原消息，再发警告**。反过来会让用户在「收到警告」到
 * 「消息被删」之间有一个窗口去截图或转发 —— 这是真实存在的对抗场景。
 */
export async function applyRuleHits(
  ctx: HitContext,
  evaluation: RuleEvaluation,
): Promise<ActionReport> {
  const { hits, timedOutRuleIds } = evaluation;
  const adminScope = ctx.scope === 'admin';

  // 超时的规则先熔断：它对任何消息都可能超时，留着就是给中继持续放血。
  if (timedOutRuleIds.length > 0) {
    const settings = await getGlobalSettings();
    for (const ruleId of timedOutRuleIds) {
      const rule = hits.find((h) => h.rule.id === ruleId)?.rule;
      if (settings.autoDisableOnRegexTimeout) {
        const reason = `匹配超时（超过 ${settings.regexTimeoutMs}ms），疑似灾难性回溯`;
        await autoDisableRule(ruleId, reason);
        logger.warn({ ruleId, ruleName: rule?.name }, '规则因匹配超时被自动停用');
      }
      if (ctx.settings.notifyAdmins && ctx.threadId !== null) {
        await sendToTopic(
          ctx.api,
          ctx.adminGroupId,
          ctx.threadId,
          `⚙️ **规则匹配超时**\n\n规则 \`${rule?.name ?? ruleId}\` 在匹配时超时，疑似存在灾难性回溯。` +
            (settings.autoDisableOnRegexTimeout ? '已自动停用以保护服务。' : '请尽快在面板中改写它。'),
        );
      }
    }
  }

  if (hits.length === 0) {
    return { outcomes: [], blocked: false, hitIds: [], violation: null, banned: false };
  }

  const actions = unionActions(hits);
  const outcomes: HitOutcome[] = [];
  let violation: ViolationOutcome | null = null;

  // 管理员的消息只审计不处罚：规则是给终端用户定的，
  // 让管理员被自己的规则静默掉是荒谬的。
  if (adminScope) {
    outcomes.push('logged_only');
    const hitIds: number[] = [];
    for (const hit of hits) {
      const id = await insertHit(ctx, hit, outcomes);
      if (id !== null) hitIds.push(id);
    }
    return { outcomes, blocked: false, hitIds, violation: null, banned: false };
  }

  // 1. 删除：用户私聊里的原消息
  if (actions.has('delete')) {
    outcomes.push((await deleteOrigin(ctx)) ? 'deleted' : 'delete_failed');
  }

  // 2. 阶梯处罚（仅 escalate 动作触发）
  if (actions.has('escalate')) {
    // 取最高分而不是求和：三条轻规则叠成一次重罚会让阈值失去意义
    const severity = hits.reduce((max, hit) => Math.max(max, hit.severity), 0);
    violation = await recordViolation({
      botId: ctx.botId,
      contactId: ctx.contact.id,
      ruleId: hits[0]?.rule.id ?? null,
      severity,
      reason: 'rule_hit',
    });
    outcomes.push(
      violation.escalated && violation.step
        ? outcomeForSanction(violation.step.type)
        : 'logged_only',
    );
  }

  // 3. 用户侧通知：显式 warn 动作，或阶梯升级（升级优先，避免一次命中发两条）
  if (violation?.escalated && violation.step) {
    if (await notifyUserOfSanction(ctx, violation, hits[0]?.rule.name ?? '')) outcomes.push('warned');
  } else if (actions.has('warn')) {
    const text = renderTemplate(ctx.settings.warnTemplate, {
      name: ctx.contact.firstName ?? '',
      username: ctx.contact.username ?? '',
      id: ctx.contact.tgUserId,
      ruleName: hits[0]?.rule.name ?? '',
      matched: hits[0]?.matchedText ?? '',
      score: violation?.score ?? ctx.contact.violationScore,
      botName: ctx.botName,
    });
    if ((await sendToUser(ctx.api, ctx.botId, ctx.contact.tgUserId, text)) !== null) {
      outcomes.push('warned');
    }
  }

  // 4. 管理员侧告警卡片
  const shouldNotify =
    ctx.settings.notifyAdmins &&
    ctx.threadId !== null &&
    (actions.has('notify') ||
      (violation?.escalated === true && violation.step?.notifyAdmins === true));

  if (shouldNotify && ctx.threadId !== null) {
    outcomes.push((await sendAlertCard(ctx, hits, outcomes, violation)) ? 'notified' : 'notify_failed');
  }

  // 5. 落审计：每条命中一行（方案里的「每条都记，动作取并集」）
  const hitIds: number[] = [];
  for (const hit of hits) {
    const id = await insertHit(ctx, hit, outcomes);
    if (id !== null) hitIds.push(id);
  }
  await bumpHitCounts(hits.map((hit) => hit.rule.id));

  // 6. 拉黑后关闭话题。刻意「关闭」而不是「删除」：删除会丢掉全部历史，
  //    而拉黑往往正是最需要保留证据的时候。管理员仍可在 Telegram 里手动删。
  let banned = false;
  if (violation?.currentType === 'ban' && ctx.topicId !== null && ctx.threadId !== null) {
    banned = true;
    await closeTopic(ctx.api, ctx.adminGroupId, {
      id: ctx.topicId,
      messageThreadId: ctx.threadId,
    });
  }

  // 7. 撤回已转发进话题的副本（先转发、后新增规则命中的场景）
  if (ctx.messageRowId !== null && actions.has('delete')) {
    const { db } = getDb();
    const rows = await db
      .select({ destChatId: messages.destChatId, relayedMessageId: messages.relayedMessageId })
      .from(messages)
      .where(eq(messages.id, ctx.messageRowId))
      .limit(1);
    const row = rows[0];
    if (row?.destChatId !== null && row?.destChatId !== undefined && row.relayedMessageId !== null) {
      await deleteMirror(ctx.api, row.destChatId, row.relayedMessageId);
    }
  }

  if (ctx.topicId !== null) await touchTopic(ctx.topicId);

  const blocked =
    violation !== null &&
    (violation.currentType === 'silence' ||
      violation.currentType === 'mute' ||
      violation.currentType === 'ban');

  return { outcomes, blocked, hitIds, violation, banned };
}

/** 按阶梯档位给用户发提示文案 */
async function notifyUserOfSanction(
  ctx: HitContext,
  violation: ViolationOutcome,
  ruleName: string,
): Promise<boolean> {
  const step = violation.step;
  if (!step) return false;

  // 静默刻意不通知：它的全部意义就是让对方无感知。
  // 一发提示等于告诉对方「你的消息被吞了，换个号再来」。
  if (step.type === 'silence') return false;

  const template =
    step.type === 'mute'
      ? ctx.settings.muteTemplate
      : step.type === 'ban'
        ? ctx.settings.banTemplate
        : ctx.settings.warnTemplate;

  const text = renderTemplate(template, {
    name: ctx.contact.firstName ?? '',
    username: ctx.contact.username ?? '',
    id: ctx.contact.tgUserId,
    ruleName,
    matched: '',
    score: violation.score,
    until: formatUntil(violation.sanction?.expiresAt ?? null),
    botName: ctx.botName,
  });

  if (!text) return false;
  return (await sendToUser(ctx.api, ctx.botId, ctx.contact.tgUserId, text)) !== null;
}

/** 话题内的告警卡片 */
async function sendAlertCard(
  ctx: HitContext,
  hits: RuleHit[],
  outcomes: HitOutcome[],
  violation: ViolationOutcome | null,
): Promise<boolean> {
  if (ctx.threadId === null) return false;

  const primary = hits[0];
  const text = renderTemplate(ctx.settings.alertCardTemplate, {
    name: [ctx.contact.firstName, ctx.contact.lastName].filter(Boolean).join(' ') || '未知用户',
    username: ctx.contact.username ? `@${ctx.contact.username}` : '（无用户名）',
    id: ctx.contact.tgUserId,
    ruleName: hits.map((h) => h.rule.name).join('、'),
    matched: primary?.matchedText ?? '',
    outcome: describeOutcomes(outcomes, violation?.currentType ?? null),
    score: violation?.score ?? ctx.contact.violationScore,
    botName: ctx.botName,
    reason: primary?.rule.note ?? '',
  });

  const id = await sendToTopic(ctx.api, ctx.adminGroupId, ctx.threadId, text);
  return id !== null;
}
